tracked/tracked.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2022-2024 Sauce Labs Inc., all rights reserved.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package tracked
+
+import (
+	"io"
+	"net"
+	"sync"
+	"sync/atomic"
+
+	"github.com/mmatczuk/connfu"
+)
+
+// ConnObserver allows to observe the number of bytes read and written from a connection.
+type ConnObserver struct {
+	rx atomic.Uint64
+	tx atomic.Uint64
+}
+
+// Rx returns the number of bytes read from the connection.
+// It requires TrackTraffic to be set to true, otherwise it returns 0.
+func (o *ConnObserver) Rx() uint64 {
+	return o.rx.Load()
+}
+
+// Tx returns the number of bytes written to the connection.
+// It requires TrackTraffic to be set to true, otherwise it returns 0.
+func (o *ConnObserver) Tx() uint64 {
+	return o.tx.Load()
+}
+
+func (o *ConnObserver) addRx(n uint64) {
+	o.rx.Add(n)
+}
+
+func (o *ConnObserver) addTx(n uint64) {
+	o.tx.Add(n)
+}
+
+type closeConn struct {
+	net.Conn
+	l closeListener // this is a field to avoid ambiguous selector error on Close method
+}
+
+func (c *closeConn) Close() error {
+	return c.l.Close()
+}
+
+type closeListener struct {
+	close   func() error
+	once    sync.Once
+	onClose func()
+}
+
+func (c *closeListener) Close() error {
+	err := c.close()
+	c.once.Do(c.onClose)
+	return err
+}
+
+// conn is a net.Conn that tracks the number of bytes read and written.
+// It needs to be configured before first use by setting TrackTraffic and onClose if needed.
+type conn struct {
+	net.Conn
+	o ConnObserver
+}
+
+func (c *conn) Read(p []byte) (n int, err error) {
+	n, err = c.Conn.Read(p)
+	c.o.addRx(uint64(n))
+	return
+}
+
+func (c *conn) Write(p []byte) (n int, err error) {
+	n, err = c.Conn.Write(p)
+	c.o.addTx(uint64(n))
+	return
+}
+
+func (c *conn) ReadFrom(r io.Reader) (n int64, err error) {
+	n, err = c.Conn.(io.ReaderFrom).ReadFrom(r)
+	c.o.addTx(uint64(n))
+	return
+}
+
+type Builder struct {
+	// TrackTraffic enables counting of bytes read and written by the connection.
+	// Use Rx and Tx to get the number of bytes read and written.
+	TrackTraffic bool
+
+	// OnClose is called after the underlying connection is closed and before the Close method returns.
+	// OnClose is called at most once.
+	OnClose func()
+}
+
+func (b Builder) Build(c net.Conn) (net.Conn, *ConnObserver) {
```

Comment:
I'm not sure if this signature is comfortable enough. Even if you wrap many connections, you only track traffic once. This is most of the time *ConnObserver is nil.

Reply:
Maybe it could provide a set of functions that build certain types.